### ISSUE TITLE
Automatic note closing on changeset close

### ISSUE THREAD

**Issue by Zaczero** [openstreetmap-ng/openstreetmap-ng#126]
I thought of an idea, some editors indicate which notes have been closed during the editing process:

![image](https://github.com/user-attachments/assets/a1a21e87-e363-4b51-9221-a27d937d49a4)

However, closing a note before uploading the changes is not practically correct and may introduce confusion.

- Let's introduce a special changeset tagging <kbd>closes:note</kbd> that will automatically close the given notes whenever that changeset is closed.
- Format will be similar to <kbd>closed:note</kbd>: numeric IDs separated by <kbd>;</kbd>.
- By default, the changeset's comment is used as the closing message on the notes (if provided).
- However, it may be possible to provide alternate group message via <kbd>closes:note:comment</kbd>.
- Individual note messages can be configured via <kbd>closes:note:ID:comment</kbd>.
- If no comment is provided on the changeset, or in the override tagging, the notes are closed without a message.
- In a case the note is already closed, do nothing.
- Testing would be nice but we can do that later (outside of the scope of the bounty).

**Comment by starsep** [comment]
noteSolver plugin for JOSM could potentially use such tags: https://github.com/kmpoppe/noteSolver

**Comment by Zaczero** [comment]
> noteSolver plugin for JOSM could potentially use such tags: https://github.com/kmpoppe/noteSolver

It's a good idea to associate changeset with the note closing messages. We will do that in a separate issue and likely store the association info in the db.

**Comment by subhu339** [comment]
Hi, is this issue still open? If yes, I’d like to work on it. Thanks!

**Comment by hobyt-aluzar** [comment]
I will work on this with a small backend-only implementation: parse the closes:note tag when a changeset is closed, resolve the default/per-note closing comments from the changeset tags, close only notes that are still open, and keep missing or already-closed notes from blocking changeset closure. I will include focused coverage for the tag parsing and close behavior.

**Comment by raxmichter** [comment]
I will take a focused backend pass on this: add parsing for the proposed `closes:note` changeset tags, close only still-open notes when the changeset is closed, support default and per-note comments, and include targeted tests for the close behavior before opening a PR.

**Comment by nexicturbo** [comment]
Submitted PR #246 for this $20 bounty: https://github.com/openstreetmap-ng/openstreetmap-ng/pull/246

It adds backend handling for `closes:note` changeset tags, default/global/per-note close comments, skip behavior for missing/already-closed notes, and focused coverage for the API close flow. I also documented the unrelated CI failures on the PR.

**Comment by danny101488** [comment]
I have a local implementation patch ready for this issue. The normal fork/PR path is still blocked on my side, so I published the patch as a temporary review artifact here:

https://github.com/danny101488/Bubio-testing/blob/main/codex-earnings/osmng/osmng-126-close-notes.patch

Summary:
- parses `closes:note` as semicolon-separated note ids
- supports default close message from changeset `comment`
- supports global `closes:note:comment`
- supports per-note `closes:note:<id>:comment`
- skips invalid, duplicate, zero, unavailable, or already-closed notes
- applies both on explicit close and inactive auto-close
- includes controller/service tests for parsing and API note-closing behavior

Local verification: `git diff --check` passed. I did not run the full suite because my local machine does not have the repo's Python 3.14/Nix test environment installed, and I avoided dependency-fetching setup on this machine.

I will open a proper PR once I have a usable fork/branch path; sharing the patch now so it is reviewable meanwhile.

**Comment by ptgram4220** [comment]
Quick maintainer summary for #126:

- The issue has multiple interest/implementation comments, but the only linked open PR I found is #246.
- #246 implements backend handling for `closes:note`, `closes:note:comment`, and per-note comments, with focused API close-flow coverage.
- #246 reports static validation commands, but says targeted pytest could not be run because the local environment was missing Bun/node_modules during app startup.
- #246 currently shows a dirty merge state, so it likely needs rebase/conflict cleanup before normal review.
- A later patch artifact was posted outside the normal PR path, so #246 still looks like the main reviewable GitHub PR.

Suggested next step: ask #246 to rebase and confirm whether the missing pytest blocker is still present; then review #246 before considering patch-only submissions.

**Comment by awer123456** [comment]
I have a local draft for #126, but I see there are already multiple open implementations. Before preparing another PR, could a maintainer confirm that a new implementation is still wanted, identify what is missing from the existing PRs, and assign the issue if appropriate?

**Comment by haotvko** [comment]
Hi @Zaczero and @nexicturbo — I’m haotvko, and I’d like to help finish #126 for a small fixed fee.

PR #246 is still open with merge conflicts. I propose updating the existing implementation against current main, resolving conflicts while preserving the original author's work, and running the focused changeset/note tests. Any remaining environment or unrelated CI blockers would be documented accurately.

Would you welcome this collaboration and allocate US$10 of the listed US$20 bounty to an accepted follow-up, with the split agreed with the original contributor? If you prefer the original author to handle the update, please let me know.

Please also confirm the current cash-payout eligibility rules, minimum amount, and transaction fees before work starts. No assignment or payment is assumed. Thanks!
